### PR TITLE
Improve cancellation in StreamPipeReader.ReadAtLeastAsync

### PR DESCRIPTION
**Pull request by manandre**
Adresses post-merge comment https://github.com/dotnet/runtime/pull/52246#discussion_r639373325

/cc @stephentoub

### CODE REVIEW

File: src/libraries/System.IO.Pipelines/tests/PipeReaderReadAtLeastAsyncTests.cs [hunk 1]
```diff
@@ -141,7 +141,7 @@ public async Task CanReadAtLeast(int bufferSize, bool bufferedRead)
         [Fact]
         public Task ReadAtLeastAsyncThrowsIfPassedCanceledCancellationToken()
         {
-            return Assert.ThrowsAsync<OperationCanceledException>(async () => await PipeReader.ReadAtLeastAsync(0, new CancellationToken(true)));
+            return Assert.ThrowsAnyAsync<OperationCanceledException>(async () => await PipeReader.ReadAtLeastAsync(0, new CancellationToken(true)));
```
Comment (stephentoub) [inline src/libraries/System.IO.Pipelines/tests/PipeReaderReadAtLeastAsyncTests.cs:144]:
```suggestion
            ValueTask<ReadResult> task = PipeReader.ReadAtLeastAsync(0, new CancellationToken(true));
            return Assert.ThrowsAnyAsync<OperationCanceledException>(async () => await task);
```

**Review by stephentoub — APPROVED** [review]

File: src/libraries/System.IO.Pipelines/tests/PipeReaderCopyToAsyncTests.cs [hunk 2]
```diff
@@ -178,7 +178,7 @@ public async Task CancelingBetweenReadsThrowsOperationCancelledException()
             Pipe.Writer.WriteEmpty(10);
             await Pipe.Writer.FlushAsync();
 
-            await Assert.ThrowsAsync<OperationCanceledException>(() => task);
+            await Assert.ThrowsAnyAsync<OperationCanceledException>(() => task);
```
Comment (davidfowl) [inline src/libraries/System.IO.Pipelines/tests/PipeReaderCopyToAsyncTests.cs:181]: ?

Comment (davidfowl) [reply src/libraries/System.IO.Pipelines/tests/PipeReaderCopyToAsyncTests.cs:181]: TaskCancelled?

**Review by davidfowl — APPROVED** [review]

File: src/libraries/System.IO.Pipelines/tests/PipeReaderReadAtLeastAsyncTests.cs [hunk 3]
```diff
@@ -141,7 +141,8 @@ public async Task CanReadAtLeast(int bufferSize, bool bufferedRead)
         [Fact]
         public Task ReadAtLeastAsyncThrowsIfPassedCanceledCancellationToken()
         {
-            return Assert.ThrowsAsync<OperationCanceledException>(async () => await PipeReader.ReadAtLeastAsync(0, new CancellationToken(true)));
+            ValueTask<ReadResult> task = PipeReader.ReadAtLeastAsync(0, new CancellationToken(canceled: true));
+            return Assert.ThrowsAnyAsync<OperationCanceledException>(async () => await task);
```
[the remark this answers is not in the document]
Comment (manandre) [reply src/libraries/System.IO.Pipelines/tests/PipeReaderReadAtLeastAsyncTests.cs]: Fixed by expecting a `TaskCanceledException`.

File: src/libraries/System.IO.Pipelines/tests/PipeReaderCopyToAsyncTests.cs (hunk 2 above)
Comment (manandre) [reply src/libraries/System.IO.Pipelines/tests/PipeReaderCopyToAsyncTests.cs]: Fixed by expecting a `TaskCanceledException`.

**Pull request merged**